packages/realm-server/pg-queue.ts

```diff
@@ -0,0 +1,408 @@
+  // 4. Remember to await workLoop.shutdown() when you're done.
+  //
+  // This is separate from the constructor so you can store your WorkLoop first,
+  // *before* the runner starts doing things.
+  run(fn: (loop: WorkLoop) => Promise<void>) {
+    this.runnerPromise = fn(this);
+  }
+
+  async shutDown(): Promise<void> {
+    log.debug(`[workloop %s] shutting down`, this.label);
+    this._shuttingDown = true;
+    this.wake();
+    await this.runnerPromise;
+    log.debug(`[workloop %s] completed shutdown`, this.label);
+  }
+
+  get shuttingDown(): boolean {
+    return this._shuttingDown;
+  }
+
+  private get waker() {
+    if (!this.internalWaker) {
+      let resolve!: () => void;
+      let promise = new Promise((r) => {
+        resolve = r;
+      }) as Promise<void>;
+      this.internalWaker = { promise, resolve };
+    }
+    return this.internalWaker;
+  }
+
+  wake() {
+    log.debug(`[workloop %s] waking up`, this.label);
+    this.waker.resolve();
+  }
+
+  async sleep() {
+    if (this.shuttingDown) {
+      return;
+    }
+    let timerPromise = new Promise((resolve) => {
+      this.timeout = setTimeout(resolve, this.pollInterval);
+    });
+    log.debug(`[workloop %s] entering promise race`, this.label);
+    await Promise.race([this.waker.promise, timerPromise]);
+    log.debug(`[workloop] leaving promise race`, this.label);
+    if (this.timeout != null) {
+      clearTimeout(this.timeout);
+    }
+    this.internalWaker = undefined;
+  }
+}
+
+export default class PgQueue implements Queue {
+  #hasStarted = false;
+  #isDestroyed = false;
+
+  private pollInterval = 10000;
+  private pgClient = new PgAdapter();
+  private handlers: Map<string, Function> = new Map();
+  private notifiers: Map<number, Deferred<any>> = new Map();
+
+  private jobRunner: WorkLoop | undefined;
+  private notificationRunner: WorkLoop | undefined;
+
+  private async query(expression: Expression) {
+    return await query(this.pgClient, expression);
+  }
+
+  private addNotifier(id: number, n: Deferred<any>) {
+    if (!this.notificationRunner && !this.#isDestroyed) {
+      this.notificationRunner = new WorkLoop(
+        'notificationRunner',
+        this.pollInterval,
+      );
+      this.notificationRunner.run(async (loop) => {
+        await this.pgClient.listen(
+          'jobs_finished',
+          loop.wake.bind(loop),
+          async () => {
+            while (!loop.shuttingDown) {
+              await this.drainNotifications(loop);
+              await loop.sleep();
+            }
+          },
+        );
+      });
+    }
+    this.notifiers.set(id, n);
+  }
+
+  private async drainNotifications(loop: WorkLoop) {
+    while (!loop.shuttingDown) {
+      let waitingIds = [...this.notifiers.keys()];
+      log.debug('jobs waiting for notification: %s', waitingIds);
+      let result = (await this.query([
+        `select id, status, result from jobs where status != 'unfulfilled' and (`,
```

Comment:
Perhaps we should format SQL with capitalized keywords like we do elsewhere in our codebase?